### ISSUE TITLE
Incorrect epoch number in TensorBoard callback when using batch-level metrics

### ISSUE THREAD

**Issue by bersbersbers** [tensorflow/tensorflow#26873]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes, see below
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10
- TensorFlow installed from (source or binary): pip install tensorflow
- TensorFlow version (use command below): `tensorflow==1.13.1` and `tf-nightly==1.14.1.dev20190318` (`v1.12.0-10310-g3db23915df 1.14.1-dev20190318`)
- Python version: 3.7.2
- Bazel version (if compiling from source): n/a
- GCC/Compiler version (if compiling from source): n/a
- CUDA/cuDNN version: n/a
- GPU model and memory: n/a

**Describe the current behavior**
When collecting batch-level metrics, epochs are incorrectly numbered with the current `_samples_seen`

**Describe the expected behavior**
When collecting batch-level metrics, epochs should be numbered with the epoch number

**Code to reproduce the issue**
```
from tensorflow import keras


class MyTensorBoard1(keras.callbacks.TensorBoard):
    def _write_custom_summaries(self, step, logs=None):
        print(logs, "at step =", step, "with regular callback")


class MyTensorBoard2(keras.callbacks.TensorBoard):
    def _write_custom_summaries(self, step, logs=None):
        print(logs, "at step =", step, "with per-batch callback")


layer = keras.layers.Input(shape=(1,))
model = keras.models.Model(inputs=layer, outputs=layer)
model.compile(optimizer='adam', loss='mse')
Callback1 = MyTensorBoard1()
Callback2 = MyTensorBoard2(update_freq='batch')
model.fit(x=list(range(6)), y=list(range(6)), epochs=10, callbacks=[Callback1, Callback2])
```

To run this code in current nightly, remove `Callback1` and focus on output of first epoch. Notice `at step = 6` in first epoch:

```
Epoch 1/10
{'batch_loss': 0.0} at step = 0 with per-batch callback
{'epoch_loss': 0.0} at step = 6 with per-batch callback
6/6 [==============================] - 0s 15ms/sample - loss: 0.0000e+00
```

**Other info / logs**
In my opinion, the bug is here:
https://github.com/tensorflow/tensorflow/blob/6612da89516247503f03ef76e974b51a434fb52e/tensorflow/python/keras/callbacks.py#L1158

and still here:
https://github.com/tensorflow/tensorflow/blob/c66b603990b9404dc1eb57de9d595aa0ffc8197f/tensorflow/python/keras/callbacks.py#L1288

In `on_epoch_end`, `step` should always be set to `epoch`.

**Comment by bersbersbers** [comment]
To explicitly include a thought of mine:

My reporting this issue is based on the expectation that `epoch_` metrics should behave the same regardless of how often `batch_` metrics are written, such that `epoch_` metrics are always comparable across runs with different `update_freq`. 

Of course, if `batch_` and `epoch_` metrics are assumed to be interpreted in conjunction for one run, there is a need to know how to interleave the two, in which case I understand the current behavior somewhat. (But then, why does `on_batch_end` use `self._total_batches_seen`, while `on_epoch_end` uses `self._samples_seen`?) In that case, however, TensorBoard should have an option to ignore the x values from summaries - this is the issue I have referenced above this comment.

**Comment by jvishnuvardhan** [comment]
@bersbersbers Could you post this in Tensorboard repo [here](https://github.com/tensorflow/tensorboard/issues). They are actively resolving the issues in Tboard repo so    it will get better visibility and faster resolution if you post it there. Close this issue when you post it on tensorboard repo. Thanks!

**Comment by bersbersbers** [comment]
@jvishnuvardhan are you certain this a TensorBoard issue? If my guesses above are correct, it's `python/keras/callbacks.py` in the TensorFlow repo which causes the issue. If you certain that TensorBoard is the correct repo to report this issue, I am happy to report it there.

**Comment by chunduriv** [comment]
@bersbersbers,

Sorry for the late response. The Epoch number is updating properly from `TF2.3` onward. Please find the [gist](https://colab.research.google.com/gist/chunduriv/c1c2fc60ed4e4c8bf878315d6ff70c87/26873.ipynb) for reference. Thanks!

**Comment by bersbersbers** [comment]
@chunduriv that's actually hard to verifiy as the gist does not print the debug output. I used the following code in TF2.8.0

```python
from tensorflow import keras
layer = keras.layers.Input(shape=(1,))
model = keras.models.Model(inputs=layer, outputs=layer)
model.compile(optimizer='adam', loss='mse')
Callback1 = keras.callbacks.TensorBoard(log_dir='logs/per_epoch', update_freq='epoch')
Callback2 = keras.callbacks.TensorBoard(log_dir='logs/per_batch', update_freq='batch')
model.fit(x=list(range(6)), y=list(range(6)), batch_size=2, epochs=10, callbacks=[Callback1, Callback2])
# then run tensorboard --logdir logs
```

and was surprised to see that the two curves are exactly identical:

![image](https://user-images.githubusercontent.com/12128514/166895986-e23fc2ab-0b7c-47aa-9074-8b2c7c83ccf4.png)

Is that intended, given that there are 10 epochs and 6 samples in 3 batches of 2 samples each?

**Comment by SuryanarayanaY** [comment]
Hi @bersbersbers,

I have gone through the above issue and would like to share few things.As per your code my understanding is that the model you created is having 'Zero' parameters as per `Model.summary()` function.Hence there is no question of updating the weights and thereby loss.Hence the loss curve is Flat irrespective of epochs.
Please note Input() is basically to create a placeholder and not exactly a layer.You have to connect input with any of the layer in order to create a Model.Hence i have created a Functional API Model for you equivalent to your code and plotted loss on tensorboard also.Please refer attached [gist](https://colab.research.google.com/gist/SuryanarayanaY/0f0466839ff35e37e5867aa55e42fb6d/26873.ipynb).
I hope iam addressing issue correctly.If not please let us know.

Thankyou!

**Comment by bersbersbers** [comment]
@SuryanarayanaY yes, you are misinterpreting the issue. This issue is not about loss values, so a model with zero parameters is perfectly fine. By contrast, this issue is about epoch values.

At the moment, I did not text the four-years-old original issue. What I can easily reproduce with TF 2.10 is what I reported last in https://github.com/tensorflow/tensorflow/issues/26873#issuecomment-1118349355: `update_freq` seems to be ignored - the outputs of both callbacks are all identical despite different values for `update_freq`.

**Comment by bersbersbers** [comment]
Related to that: https://github.com/keras-team/keras/issues/16173

**Comment by SuryanarayanaY** [comment]
Thanks @bersbersbers ,
We will track the issue from Keras Repo.We Will keep you informed once the issue resolve.
Thankyou!

**Comment by SuryanarayanaY** [comment]
Hi @bersbersbers ,
As this issue is currently active on Keras-repo,could you please close it here so that we can track the issue there at Keras-repo ?

Thankyou in advance.

**Comment by bersbersbers** [comment]
@SuryanarayanaY I am not sure https://github.com/keras-team/keras/issues/16173 is a good replacement for this issue, as that issue tracks bringing back a functionality that was deleted; this issue here is about a bug within that functionality. Having said that, this issue is four years old and there is a good chance it is outdated if https://github.com/keras-team/keras/issues/16173 is closed one way or the other. I will close this and simply reopen once I run across this issue again (which is unlikely considering that I am not pursueing deep learning at this point in time).
